pg_shard--1.1.sql

```diff
@@ -140,3 +140,51 @@ $sync_table_metadata_to_citus$ LANGUAGE 'plpgsql';
 
 COMMENT ON FUNCTION sync_table_metadata_to_citus(text)
 		IS 'synchronize a distributed table''s pg_shard metadata to CitusDB';
+
+-- Prepares a specified distributed table to work with COPY operations by
+-- creating a temporary table and trigger to handle rows coming from a COPY.
+-- After each INSERT, a provided sequence is incremented to track the total -- number of copied rows.
+CREATE FUNCTION prepare_distributed_table_for_copy(relation regclass, sequence regclass) RETURNS void
+AS $pdtfc$
+	DECLARE
+		table_name text;
+		temp_table_name text;
+		attr_names text[];
+		attr_list text;
+		param_list text;
+		using_list text;
+		insert_command text;
+		func_tmpl CONSTANT text  := $$  CREATE FUNCTION pg_temp.copy_to_insert() RETURNS trigger
+										AS $cti$
+											BEGIN
+												EXECUTE %L USING %s;
+												PERFORM nextval(%L);
+												RETURN NULL;
+											END;
+										$cti$ LANGUAGE plpgsql VOLATILE;
+									$$;
+		table_tmpl CONSTANT text := 'CREATE TEMPORARY TABLE %I (LIKE %s)';
+		trg_tmpl CONSTANT text := $$CREATE TRIGGER copy_to_insert
+									BEFORE INSERT ON %s
+									FOR EACH ROW EXECUTE PROCEDURE pg_temp.copy_to_insert()$$;
+	BEGIN
+		SELECT relname INTO STRICT table_name FROM pg_class WHERE oid = relation;
+		temp_table_name = format('%s_copy_facade', table_name);
+
+		SELECT array_agg(attname) INTO STRICT attr_names FROM pg_attribute WHERE attrelid = relation AND
```

Comment:
This means the `INSERT` trigger will generate a statement with _all_ user-visible attributes fully specified. If the overarching `COPY` does not specify all columns, one of two things could happen:

  * If `NEW.missing_col` is provided as `NULL`, we'd explicitly include a `NULL` in the `INSERT`. This could be wrong if there is a default value specified
  * If `NEW.missing_col` already replaces unspecified columns with their default values, the trigger would just explicitly specify the default value, which would be correct.

The copy shell script presently doesn't allow omitting columns: it's expected the CSV or whatnot has the same dimensions as the table. So this is kind of a moot point… I'll check about the `NULL` vs default value distinction.